src/adapter/src/coord/cluster_scheduling.rs

```diff
@@ -0,0 +1,169 @@
+                                {
+                                    if mv.refresh_schedule.is_some() {
+                                        Some(id)
+                                    } else {
+                                        None
+                                    }
+                                } else {
+                                    None
+                                }
+                            })
+                            .collect_vec();
+
+                        // Check whether there is at least one REFRESH MV needing a refresh.
+                        let local_read_ts = self.get_local_read_ts().await;
+                        decisions.push((
+                            cluster.id,
+                            refresh_mvs.into_iter().any(|mv| {
+                                let mv_frontier = &self
+                                    .controller
+                                    .storage
+                                    .collection(*mv)
+                                    .expect("the storage controller should know about MVs that exist in the catalog")
+                                    .write_frontier;
+                                mv_frontier.less_than(&local_read_ts)
+                            }),
+                        ));
+                    }
+                }
+            }
+        }
+        decisions
+    }
+
+    #[mz_ore::instrument(level = "debug")]
+    pub(crate) async fn handle_scheduling_decisions(
+        &mut self,
+        decisions: Vec<(&'static str, Vec<(ClusterId, bool)>)>,
+    ) {
+        for (policy_name, decisions) in decisions.iter() {
+            for (cluster_id, decision) in decisions {
+                self.scheduling_decisions
+                    .entry(*cluster_id)
+                    .or_insert_with(Default::default)
+                    .insert(policy_name, *decision);
+            }
+        }
+
+        // Clean up those clusters from `scheduling_decisions` that
+        // - have been dropped, or
+        // - were switched to unmanaged, or
+        // - were switched to `SCHEDULE = MANUAL`.
+        for cluster_id in self.scheduling_decisions.keys().cloned().collect_vec() {
+            match self.get_managed_cluster_config(cluster_id) {
+                None => {
+                    // Cluster have been dropped or switched to unmanaged.
+                    self.scheduling_decisions.remove(&cluster_id);
+                }
+                Some(managed_config) => {
+                    if matches!(managed_config.schedule, ClusterScheduleOptionValue::Manual) {
+                        self.scheduling_decisions.remove(&cluster_id);
+                    }
+                }
+            }
+        }
+
+        for (cluster_id, decisions) in self.scheduling_decisions.clone() {
+            // If all policies have made a decision about this cluster
+            if POLICIES.iter().all(|policy| decisions.contains_key(policy)) {
```

Comment:
I didn't follow the explanation of why we need to wait for all policies to run at least once before taking the first action. Execution of policies is inherently asynchronous, and it doesn't seem like there's anything particularly special about waiting for each policy to run at least once?

Reply:
Ah, I see. Ok, I think that makes sense. I do think calling out here that the goal is to ensure all policies have a chance to run at least once after a restart would be helpful!

Reply:
Done, I've added the following comment:
```
// We touch a cluster only when all policies have made a decision about it. This is
// to ensure that after an envd restart all policies have a chance to run at least once
// before we turn off a cluster, to avoid spuriously turning off a cluster and possibly
// losing a hydrated state.
```